### PR TITLE
[DEVX-210]: YAML based workflow creation

### PR DESCRIPTION
**Pull request by sainivedh**
## Why
 - Enable users to create workflow with a simple YAML config script. 
 - Also helps user to export existing workflow as a YAML

## How
- Developed based on Dali team [yaml2workflow](https://github.com/isaac-chung/yaml2workflow) library
- Design Doc - [YAML Workflow Creation in SDK](https://clarifai.atlassian.net/wiki/spaces/DevX/pages/3267035313/Yaml+Workflow+Creation+in+SDK)
### Flow Chart of Node Creation
image

#### The Workflow Create function allows you to create a new workflow, specified by a yaml config file.
```python
# Note: CLARIFAI_PAT must be set as env variable.
from clarifai.client.app import App
app = App(app_id="app_id", user_id="user_id")
workflow = app.create_workflow(config_filepath="config.yml")
```

#### The Workflow Export function allows you to export an existing workflow from Clarifai and save it to a local file for future use.
```python
# Note: CLARIFAI_PAT must be set as env variable.
from clarifai.client.workflow import Workflow
workflow = Workflow("https://clarifai.com/clarifai/main/workflows/Demographics")
workflow.export('demographics_workflow.yml')
```

## What
- Added Workflow Validate, Export, Create functionalities
- Added list module versions
- Added display of workflow graph 
```yaml
workflow:
  id: test-wf-prompter
  nodes:
    - id: prompter
      model:
          model_id: prompter
          model_type_id: prompter
          description: 'Sentiment Analysis'
          output_info:
            params:
              prompt_template: 'Classify sentiment between postive and negative {data.text.raw}'

    - id: llm
      model:
          user_id: openai
          model_id: GPT-3_5-turbo
          app_id: chat-completion

      node_inputs:
        - node_id: prompter
```
image

## Tests
 - Reused the same test cases as in yaml2workflow for export, create, validate
 - Create covers multiple scenarios as multi branch/multi node/custom model/public model...

### CODE REVIEW

**Review by chungisaac — CHANGES_REQUESTED** [review]

File: clarifai/workflows/validate.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+_data_schema = Schema({
+    "workflow": {
+        "id":
+            _id_validator,
+        "nodes":
+            And(
+                len,
+                [{
+                    "id":
+                        And(str, len),  # Node IDs are not validated as IDs by the API.
```
[the remark this answers is not in the document]
Comment (sainivedh) [reply clarifai/workflows/validate.py:43]: Yes, Used this [pre-commit config](https://github.com/Clarifai/clarifai-python/blob/master/.pre-commit-config.yaml).

File: clarifai/client/app.py [hunk 2]
```diff
@@ -236,30 +242,93 @@ def create_model(self, model_id: str, **kwargs) -> Model:
 
     return Model(model_id=model_id, **kwargs)
 
-  def create_workflow(self, workflow_id: str, **kwargs) -> Workflow:
+  def create_workflow(self, config_filepath: str, generate_new_id: bool = False) -> Workflow:
```
Comment (sanjaychelliah) [inline clarifai/client/app.py:245]: can this arg be `generate_random_id` instead of `generate_new_id`? WDYT?

Comment (chungisaac) [reply clarifai/client/app.py:245]: it's definitely closer to the actual implementation, so I'm for it as well.

File: clarifai/client/app.py [hunk 3]
```diff
@@ -236,30 +242,93 @@ def create_model(self, model_id: str, **kwargs) -> Model:
...
     if response.status.code != status_code_pb2.SUCCESS:
       raise Exception(response.status)
     self.logger.info("\nWorkflow created\n%s", response.status)
-    kwargs.update({'app_id': self.id, 'user_id': self.user_id})
 
-    return Workflow(workflow_id=workflow_id, **kwargs)
+    dict_response = MessageToDict(response, preserving_proto_field_name=True)
+    # Display the workflow nodes tree.
+    display_workflow_tree(dict_response["workflows"][0]["nodes"])
+    kwargs = self.process_response_keys(dict_response[list(dict_response.keys())[1]][0],
```
[the remark this answers is not in the document]
Comment (sainivedh) [reply clarifai/client/app.py:328]: Yes, plan is to make it a CLI feature. This is where a unified Search helps(`GetWorkflow`). For now, as its always on should we provide a toggle argument

File: clarifai/client/app.py (hunk 2 above)
Comment (sainivedh) [reply clarifai/client/app.py:245]: Even thought it just generates random id. I think `generate_new_id` is more intuitive

File: clarifai/workflows/validate.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
+from schema import And, Optional, Regex, Schema, SchemaError, Use
+
+# Non-empty, up to 32-character ASCII strings with internal dashes and underscores.
+_id_validator = And(str, lambda s: 0 < len(s) <= 32, Regex(r'^[0-9A-Za-z]+([-_][0-9A-Za-z]+)*$'))
+
+# 32-character hex string, converted to lower-case.
+_hex_id_validator = And(str, Use(str.lower), Regex(r'^[0-9a-f]{32}'))
+
+
+def _model_does_not_have_model_version_id_and_other_fields(m):
```
Comment (sanjaychelliah) [inline clarifai/workflows/validate.py:10]: Can we name the function anything simpler rather than the whole Description?

**Review by sanjaychelliah — COMMENTED** [review]
LGTM

File: clarifai/workflows/validate.py (hunk 4 above)
Comment (sainivedh) [reply clarifai/workflows/validate.py:10]: Good Point. But I feel in `validate.py` the long name is needed so as to be more specific on functionality. However we can make function names in `workflows/utils.py` much shorter. Let me do that.

File: tests/workflow/test_create.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+import logging
+import os
+import typing
+
+import pytest
+
+from clarifai.client.user import User
+
+CREATE_APP_USER_ID = os.environ["CLARIFAI_USER_ID"]
+CREATE_APP_ID = "ci_workflow_create"
```
[the remark this answers is not in the document]
Comment (sainivedh) [reply tests/workflow/test_create.py]: Tried this. Tests are failing except for MacOS, not sure of the reason. Reverted back
https://github.com/Clarifai/clarifai-python/actions/runs/6270166647

File: clarifai/utils/logging.py [hunk 6]
```diff
@@ -1,16 +1,66 @@
...
 from rich.table import Table
 from rich.traceback import install
+from rich.tree import Tree
 
 install()
 
 
-def table_from_dict(data, column_names, title="") -> Table:
+def display_workflow_tree(nodes_data: List[Dict]) -> None:
+  """Displays a tree of the workflow nodes."""
```
[the remark this answers is not in the document]
Comment (sainivedh) [reply clarifai/utils/logging.py:15]: Added `build_node_tree` tests in `tests/workflow/test_nodes_display.py`.

File: README.md [hunk 7]
```diff
@@ -195,7 +195,23 @@ all_workflow = app.list_workflow()
 # List all workflow in community filtered by description
 all_face_community_workflows = App().list_workflows(filter_by={"query": "face"}, only_in_app=False) # Get all face related workflows
 ```
+#### Workflow Create
+#### The Workflow Create function allows you to create a new workflow, specified by a yaml config file.
```
[the remark this answers is not in the document]
Comment (chungisaac) [reply README.md]: Still has the header hashtags before `Create`. Same for the other line.

[the remark this answers is not in the document]
Comment (sainivedh) [reply README.md]: Done

File: clarifai/client/app.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (chungisaac) [reply clarifai/client/app.py:327]: Yes we could provide a toggle arg. Maybe like `display=True` as default

**Pull request merged**